### PR TITLE
Bug 2000352: [CORS-1716] vsphere: set the imported ova hardware version

### PR DESCRIPTION
**Pull request by jcpowermac**
The out-of-tree csi requires hardware 15, vSphere 6.7U3 and greater

Hardware 15 is only supported on 6.7 U2 and greater.
This PR checks for each ESXi host version, if less than 6.7 U3
sets the hardware version to 13. If greater than or equal to 6.7U3
the hardware version is set to 15.

- [x] See if its possible to simplify the version logic

### CODE REVIEW

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 1]
```diff
@@ -189,11 +200,45 @@ func findImportOvaParams(client *vim25.Client, datacenter, cluster, datastore, n
...
 
+		// If HardwareVersion is 13 there is no reason to continue checking
+		// There is a ESXi host that does not support hardware 15.
+		if importOvaParams.HardwareVersion != "vmx-13" {
+			esxiHostVersion, err := version.NewVersion(hostSystemManagedObject.Config.Product.Version)
+			if err != nil {
+				return nil, err
+			}
+			// Check if ESXi host is greater than or equal to 6.7
+			if constraints.Check(esxiHostVersion) {
```
Comment (jcpowermac) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go:216]: The logic that I have here for constraints needs to be refined. Wanted to get a PR open to test with CI

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 2]
```diff
@@ -375,6 +425,19 @@ func resourceVSpherePrivateImportOvaCreate(d *schema.ResourceData, meta interfac
...
 	log.Printf("[DEBUG] %s: mark as template", vm.Name())
 
+	task, err := vm.UpgradeVM(ctx, importOvaParams.HardwareVersion)
+
+	if err != nil {
+		return errors.Errorf("failed to upgrade vm to: %s, %s", importOvaParams.HardwareVersion, err)
+
+	}
+
+	err = task.Wait(ctx)
```
Comment (jcpowermac) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go:435]: Need to review if this `Wait()` is correct

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 3]
```diff
@@ -189,11 +198,47 @@ func findImportOvaParams(client *vim25.Client, datacenter, cluster, datastore, n
...
 		foundDatastore := false
 		foundNetwork := false
-		err := hostObj.Properties(ctx, hostObj.Reference(), []string{"network", "datastore", "runtime"}, &hostSystemManagedObject)
+		err := hostObj.Properties(ctx, hostObj.Reference(), []string{"config.product", "network", "datastore", "runtime"}, &hostSystemManagedObject)
 		if err != nil {
 			return nil, err
 		}
 
+		// If HardwareVersion is 13 there is no reason to continue checking
+		// There is a ESXi host that does not support hardware 15.
```
Comment (e-tienne) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go:207]: I'm not sure to understand that comment. Are we talking about all the older [ESXi hosts 6.7 and earlier not supporting Hardware version 15](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.vm_admin.doc/GUID-789C3913-1053-4850-A0F0-E29C3D32B6DA.html)?

**Review by e-tienne — COMMENTED** [review]
Looks good to me in general; do we have a card for this?
Do you know if we need to do additional testing when bumping hardware Versions? I'm wondering if there is any risk at all.

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 4]
```diff
@@ -189,11 +198,45 @@ func findImportOvaParams(client *vim25.Client, datacenter, cluster, datastore, n
...
+				if build >= 14320388 {
+					importOvaParams.HardwareVersion = "vmx-15"
+				} else {
+					importOvaParams.HardwareVersion = "vmx-13"
+				}
+
+			} else if esxiHostVersion.GreaterThan(v67) {
+				importOvaParams.HardwareVersion = "vmx-15"
+			} else {
+				importOvaParams.HardwareVersion = "vmx-13"
```
Comment (patrickdillon) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go:236]: can we set all of these `"vmx-13"` and ``"vmx-15"`s to constant variables

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 5]
```diff
@@ -375,6 +423,21 @@ func resourceVSpherePrivateImportOvaCreate(d *schema.ResourceData, meta interfac
 	}
 	log.Printf("[DEBUG] %s: mark as template", vm.Name())
 
+	task, err := vm.UpgradeVM(ctx, importOvaParams.HardwareVersion)
```
Comment (patrickdillon) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go:426]: do we need more context here? like a comment or something in the commit message? I'm not sure what this is doing or why.

**Review by patrickdillon — COMMENTED** [review]
In general this looks good to me, but I am definitely not a subject matter expert. 

Left a few nitpicky style comments. 

Would like to see e2e-vsphere install then I'm happy to lgtm.

/approve

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 6]
```diff
@@ -189,11 +204,45 @@ func findImportOvaParams(client *vim25.Client, datacenter, cluster, datastore, n
...
 		}
 
+		// If HardwareVersion is 13 there is no reason to continue checking
+		// There is a ESXi host that does not support hardware 15.
+		if importOvaParams.HardwareVersion != vmx13 {
+			esxiHostVersion, err := version.NewVersion(hostSystemManagedObject.Config.Product.Version)
+			if err != nil {
+				return nil, err
+			}
+
```
Comment (rna-afk) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go:219]: Would this be better?
```
importOvaParams.HardwareVersion = vmx13
if esxiHostVersion.Equal(v67) {
	build, err := strconv.Atoi(hostSystemManagedObject.Config.Product.Build)
	if err != nil {
		return nil, err
	}
	// This is the ESXi 6.7 U3 build number
	// Anything less than this version is unsupported with the
	// out-of-tree CSI.
	// https://kb.vmware.com/s/article/2143838
	// https://vsphere-csi-driver.sigs.k8s.io/supported_features_matrix.html
	if build >= Esxi67U3BuildNumber {
		importOvaParams.HardwareVersion = vmx15
	}
} else if esxiHostVersion.GreaterThan(v67) {
	importOvaParams.HardwareVersion = vmx15
}
```

Comment (jcpowermac) [reply pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go:219]: @rna-afk thanks for the suggestion, yes so much better with less branches

**Comment by jcpowermac** [comment]
Issue with CI infra...
/test e2e-vsphere

**Comment by jcpowermac** [comment]
Individual test failure, unrelated to change.

Based on logging template was hardware version 15.
```
time="2021-09-02T13:51:57Z" level=debug msg="2021-09-02T13:51:57.674Z [DEBUG] plugin.terraform-provider-vsphere: 2021/09/02 13:51:57 [DEBUG] Found current hardware version: 15"
time="2021-09-02T13:51:57Z" level=debug msg="2021-09-02T13:51:57.674Z [DEBUG] plugin.terraform-provider-vsphere: 2021/09/02 13:51:57 [DEBUG] Powering on virtual machine \"/SDDC-Datacenter/vm/ci-op-j5tf7xp4-67021-9k2n9/ci-op-j5tf7xp4-67021-9k2n9-bootstrap\""
time="2021-09-02T13:51:57Z" level=debug msg="2021-09-02T13:51:57.674Z [DEBUG] plugin.terraform-provider-vsphere: 2021/09/02 13:51:57 [DEBUG] blockUntilReadyForMethod: Going to block until \"PowerOnVM_Task\" is no longer in the Disabled Methods list for vm vm-504897"
time="2021-09-02T13:51:57Z" level=debug msg="2021-09-02T13:51:57.674Z [DEBUG] plugin.terraform-provider-vsphere: 2021/09/02 13:51:57 [DEBUG] Fetching properties for VM \"/SDDC-Datacenter/vm/ci-op-j5tf7xp4-67021-9k2n9/ci-op-j5tf7xp4-67021-9k2n9-bootstrap\""
time="2021-09-02T13:51:57Z" level=debug msg="2021-09-02T13:51:57.721Z [DEBUG] plugin.terraform-provider-vsphere: 2021/09/02 13:51:57 [DEBUG] Fetching properties for VM \"/SDDC-Datacenter/vm/ci-op-j5tf7xp4-67021-9k2n9/ci-op-j5tf7xp4-67021-9k2n9-master-0\""
time="2021-09-02T13:51:57Z" level=debug msg="2021-09-02T13:51:57.726Z [DEBUG] plugin.terraform-provider-vsphere: 2021/09/02 13:51:57 [DEBUG] VM \"/SDDC-Datacenter/vm/ci-op-j5tf7xp4-67021-9k2n9/ci-op-j5tf7xp4-67021-9k2n9-master-0\" is powered off, attempting to power on."
```

**Pull request merged**